service/matching/matching_engine.go

```diff
@@ -371,60 +371,129 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
+
+	e.partitionsLock.RLock()
+	pm, ok := e.partitions[key]
+	e.partitionsLock.RUnlock()
+
+	if ok {
+		return pm, nil
+	}
+
+	if !create {
+		return nil, nil
+	}
+
+	pm, created, err := e.createTaskQueuePartitionManager(partition, loadCause)
 	if err != nil || pm == nil {
 		return nil, err
 	}
+
+	if pm.Partition().IsRoot() && created {
+		// Whenever a root partition is loaded we need to force all other partitions to load.
+		// If there is a backlog of tasks on any child partitions force loading will ensure that they
+		// can forward their tasks the poller which caused the root partition to be loaded.
+		// These partitions could be managed by this matchingEngineImpl, but are most likely not.
+		// We skip checking and just make gRPC requests to force loading them all.
+		err := e.forceLoadAllNonRootPartitions(ctx, partition)
+		if err != nil {
+			e.logger.Error("Failed to force load all non-root partitions after root partition was loaded",
+				tag.WorkflowNamespaceID(partition.NamespaceId().String()),
+				tag.WorkflowTaskQueueName(partition.TaskQueue().Name()),
+				tag.WorkflowTaskQueueType(partition.TaskType()),
+				tag.Error(err))
+		}
+	}
+
 	if err = pm.WaitUntilInitialized(ctx); err != nil {
 		e.unloadTaskQueuePartition(pm, unloadCauseInitError)
 		return nil, err
 	}
 	return pm, nil
 }
 
-// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
-// to get new range from DB and create one. This does not block for the task queue to be initialized.
-func (e *matchingEngineImpl) getTaskQueuePartitionManagerNoWait(
+func (e *matchingEngineImpl) forceLoadAllNonRootPartitions(
+	ctx context.Context,
+	rootPar tqid.Partition,
+) error {
+	namespaceId := rootPar.NamespaceId()
+
+	namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(namespaceId)
+	if err != nil {
+		return err
+	}
+
+	nsName := namespaceEntry.Name()
+	taskQueue := rootPar.TaskQueue()
+
+	partitionTotal := e.config.NumTaskqueueReadPartitions(nsName.String(), taskQueue.Name(), taskQueue.TaskType())
+
+	for partitionId := 1; partitionId < partitionTotal; partitionId++ {
+
+		go func() {
+			resp, err := e.matchingRawClient.ForceLoadTaskQueue(ctx, &matchingservice.ForceLoadTaskQueueRequest{
+				NamespaceId: namespaceId.String(),
+				TaskQueuePartition: &taskqueuespb.TaskQueuePartition{
+					TaskQueue:     taskQueue.Name(),
+					TaskQueueType: taskQueue.TaskType(),
+					PartitionId:   &taskqueuespb.TaskQueuePartition_NormalPartitionId{NormalPartitionId: int32(partitionId)},
+				},
+			})
+			if err != nil {
+				e.logger.Error("Failed to force load non-root partition after root partition was loaded",
+					tag.WorkflowNamespaceID(nsName.String()),
+					tag.WorkflowTaskQueueName(taskQueue.Name()),
+					tag.WorkflowTaskQueueType(taskQueue.TaskType()),
+					tag.Error(err))
+				return
+			}
+
+			if !resp.WasUnloaded {
+				// Generally we should be avoiding trying to force load partitions which are already loaded.
```

Comment:
why? that seems like a quite common situation that'll happen

Reply:
My understanding is that the most common case is that all partitions are unloaded, except for the forwarding partition. We should avoid attempting to forceLoad a loaded partition because its an unnecessary rpc.

Reply:
I'll change the comment to better illustrate the situation then.